### ISSUE TITLE
Reformat support for subclass and suborder

### ISSUE THREAD

**Issue by Username-felix-is-not-available** [shenwei356/taxonkit#107]
### Prerequisites

- [x] make sure you're are using the latest version by `taxonkit version`
- [x] read the [usage](http://bioinf.shenwei.me/taxonkit/usage/)

### Describe your issue
I am working with [Kraken2](https://github.com/DerrickWood/kraken2) and the RDP database. The RDP database has a similar format to the NCBI taxonomy with a names.dmp file and a nodes.dmp file. Note, however, that the RDP IDs are custom IDs not based on the NCBI taxonomy.  RDP includes two ranks which are currently not supported by TaxonKit: subclass and suborder. Would it be possible to support these ranks in the future? This issue also applies to the SILVA and the "standard" database which contain even more ranks (not needed for me). Maybe the --format syntax could be extended to support arbitrary ranks? Users could then enter the full rank name instead of a placeholder to force TaxonKit to look for this rank.

### Example
I attached the names.dmp and nodes.dmp files for the RDP database.
[rdp_taxonomy.zip](https://github.com/user-attachments/files/17680425/rdp_taxonomy.zip)

Currently, the output of
`echo "3535" | taxonkit reformat -I 1 -f "{k};{p};{c};{o};{f};{g};{s}" --data-dir rdp_taxonomy/`
is
3535	Bacteria;"Chloroflexi";Thermomicrobia;Sphaerobacterales;Sphaerobacteraceae;Sphaerobacter;

I would like to be able to execute
`echo "3535" | taxonkit reformat -I 1 -f "{k};{p};{c};{sc};{o};{so};{f};{g};{s}" --data-dir rdp_taxonomy/`
or
`echo "3535" | taxonkit reformat -I 1 -f "{k};{p};{c};{subclass};{o};{suborder};{f};{g};{s}" --data-dir rdp_taxonomy/`
and get this output
3535	Bacteria;"Chloroflexi";Thermomicrobia;**Sphaerobacteridae**;Sphaerobacterales;**"Sphaerobacterineae"**;Sphaerobacteraceae;Sphaerobacter;

**Comment by shenwei356** [comment]
> Users could then enter the full rank name instead of a placeholder to force TaxonKit to look for this rank.

Good advice. Will try to do it next week.

**Comment by Username-felix-is-not-available** [comment]
That would be awesome!
Have a nice weekend,
Felix

**Comment by Username-felix-is-not-available** [comment]
Awesome! Thank you for the quick update, @shenwei356 . The reformat2 function does exactly what I was hoping for. Today, I tested it in depth and could not locate any issues.
Nevertheless, may I ask you to make the -T (--trim) flag from reformat also available in reformat2? Currently, I am working around that with some small Perl script, but I think it would be useful to have it directly available in TaxonKit.

**Comment by shenwei356** [comment]
<s>Please give an example of how -T should work.</s> I remember now.

E.g., for genus Escherichia (561), both `reformat` and `reformat2` works fine, with an empty species name in the end. And there's no need to use -T for `reformat`, unless you set another replacement for missing ranks.

```
$ echo 561 | taxonkit reformat -I 1
561     Bacteria;Pseudomonadota;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;

$ echo 561 | taxonkit reformat -I 1 -r unknown -T
561     Bacteria;Pseudomonadota;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;

$ echo 561 | taxonkit reformat2 
561     Bacteria;Pseudomonadota;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;
```

OK, I see. But for the current implementation. What I can do is just replacing "unknown"s in the end to empty. It should work. Will do it tomorrow.

```
$ echo 561 | taxonkit reformat2 -r unknown
561     Bacteria;Pseudomonadota;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;unknown
```

**Comment by Username-felix-is-not-available** [comment]
Yes, I am using the replacement option (-r) to set every rank with an unknown taxon name to "0". So currently, I am getting:
```
$ echo "561" | taxonkit reformat2 -I 1 -r "0" -f "{superkingdom|kingdom};{phylum};{class};{subclass};{order};{suborder};{family};{genus};{species}"
561	Bacteria;Pseudomonadota;Gammaproteobacteria;0;Enterobacterales;0;Enterobacteriaceae;Escherichia;0
```
Since "561" is the ID for a taxon at the rank genus, I would like to leave the species empty.
```
$ echo "561" | taxonkit reformat2 -I 1 -r "0" -f "{superkingdom|kingdom};{phylum};{class};{subclass};{order};{suborder};{family};{genus};{species}" -T
561	Bacteria;Pseudomonadota;Gammaproteobacteria;0;Enterobacterales;0;Enterobacteriaceae;Escherichia;
```

> What I can do is just replacing "unknown"s in the end to empty. It should work.

I think it is a little bit more tricky than just stripping off the "0"s at the end, because trim should only remove the "0"s for ranks below the rank identified by the tax ID. In the upper example "561" identifies a genus. That's why the "0" at species level should be removed. However, if I ask TaxonKit for the lineage from superkingdom to subclass, the trailing "0" should be kept (subclass > genus).
```
$ echo "561" | taxonkit reformat2 -I 1 -r "0" -f "{superkingdom|kingdom};{phylum};{class};{subclass}" -T
561	Bacteria;Pseudomonadota;Gammaproteobacteria;0
```

This is also how the original reformat is handling the trimming for the classical ranks ("102403" is another genus).

```
$ echo "102403" | taxonkit reformat -I 1 -r "0" -T -f "{k};{p};{c};{o};{f};{g};{s}"
102403	Eukaryota;Mollusca;Bivalvia;0;Poromyidae;Tropidomya;
$ echo "102403" | taxonkit reformat -I 1 -r "0" -T -f "{k};{p};{c};{o}"
102403	Eukaryota;Mollusca;Bivalvia;0
```

Sorry for causing that much trouble! :) I really appreciate your help.

**Comment by shenwei356** [comment]
OK, done. 

> I think it is a little bit more tricky than just stripping off the "0"s at the end
Agreed. They are trimmed only after seeing the lowest known rank of the taxid. 
And thanks for the example. 

- [taxonkit_darwin_arm64.tar.gz](https://github.com/user-attachments/files/17740850/taxonkit_darwin_arm64.tar.gz)
- [taxonkit_linux_amd64.tar.gz](https://github.com/user-attachments/files/17740852/taxonkit_linux_amd64.tar.gz)
- [Uploading taxonkit_windows_amd64.exe.tar.gz…]()

    $ echo 561 | taxonkit reformat2 -r unknown
    561     Bacteria;Pseudomonadota;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;unknown

    $ echo 561 | taxonkit reformat2 -r unknown -T
    561     Bacteria;Pseudomonadota;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;


    # -----------------------------------------------------------
    # another example where the order rank is missing

    $ echo 102403 | taxonkit reformat2 -I 1 -r "0" -f "{superkingdom};{phylum};{class};{order};{family};{genus};{species}"
    102403  Eukaryota;Mollusca;Bivalvia;0;Poromyidae;Tropidomya;0

    $ echo 102403 | taxonkit reformat2 -I 1 -r "0" -f "{superkingdom};{phylum};{class};{order};{family};{genus};{species}" -T
    102403  Eukaryota;Mollusca;Bivalvia;0;Poromyidae;Tropidomya;

    # and now, the lowest rank in the output format is order, but the tailing "0" is not trimmed.

    $ echo 102403 | taxonkit reformat2 -I 1 -r "0" -f "{superkingdom};{phylum};{class};{order}"
    102403  Eukaryota;Mollusca;Bivalvia;0

    $ echo 102403 | taxonkit reformat2 -I 1 -r "0" -f "{superkingdom};{phylum};{class};{order}" -T
    102403  Eukaryota;Mollusca;Bivalvia;0

**Comment by Username-felix-is-not-available** [comment]
Works like a charm! Thanks a lot, @shenwei356 .

**Comment by shenwei356** [comment]
Cheers!
